File: codes/C/Chapter 5/Example_5.3.c
#include "Example_5.3.h"

void MaximalInducedSubgraph(int** adj, int n, int k, int* result)
{
    int degree[n];
    int queue[n];
    int len = 0;
    for(int i = 0 ; i < n ; ++i){
        result[i] = 1;
        degree[i] = 0;
        for(int j = 0 ; j < n ; ++j)
            degree[i] += adj[i][j];
    }
    for(int i = 0 ; i < n ; ++i){
        if(degree[i] < k){
            result[i] = 0;
            queue[len++] = i;
        }
    }
    int head = 0;
    while(head < len){
        int u = queue[head++];
        for(int v = 0 ; v < n ; ++v){
            if(adj[u][v] && result[v]){
                degree[v]--;
                if(degree[v] < k){
                    result[v] = 0;
                    queue[len++] = v;
                }
            }
        }
    }
}

File: codes/python/Chapter 5/Example5.3.py
# !/usr/bin/python
# -*- coding:utf-8 -*-
# Copyright 2026 Xinyi
# Author  : Xinyi Hu
# Contact : samaritanhu@gmail.com
# Time    : 2026/07/21 10:00

import queue

def MaximalInducedSubgraph(adj, n, k):
    '''
    adj: adjacency list, adj[i] is an iterable of the neighbors of vertex i (0-indexed)
    n: number of vertices
    k: every remaining vertex must have degree >= k in the induced subgraph

    Returns the set U of vertices (0-indexed) of the maximal induced subgraph
    H = (U, F) such that every vertex in H has degree >= k. U is empty if no
    such non-empty subgraph exists.
    '''
    degree = [len(adj[i]) for i in range(n)]
    removed = [False] * n
    Queue = queue.Queue()
    for i in range(n):
        if degree[i] < k:
            removed[i] = True
            Queue.put(i)
    while not Queue.empty():
        u = Queue.get()
        for v in adj[u]:
            if not removed[v]:
                degree[v] -= 1
                if degree[v] < k:
                    removed[v] = True
                    Queue.put(v)
    return {i for i in range(n) if not removed[i]}
